database.cc

```diff
@@ -588,10 +580,7 @@ future<> database::parse_system_tables(distributed<service::storage_proxy>& prox
     }).then([&proxy, this] {
         return do_parse_schema_tables(proxy, db::schema_tables::TYPES, [this, &proxy] (schema_result_value_type &v) {
             auto& ks = this->find_keyspace(v.first);
-            auto&& user_types = create_types_from_schema_partition(*ks.metadata(), v.second);
-            for (auto&& type : user_types) {
-                ks.add_user_type(type);
-            }
+            create_types_from_schema_partition(*ks.metadata(), v.second);
```

Comment:
This comment is not directly targeted at this patch, but rather at 4f1260f3e36210e178869739c7e8557d038bb1da which started this. I am not sure that coupling type parsing with insertion into the keyspace is a good direction. It means that the operation which involves parsing cannot be safely rolled-back if it fails mid-way. There is no way to obtain a type instance without adding it to the keyspace.

Reply:
I think it could be changed to copy user_types_metadata, in which case this patch could not depend on build() adding the types. Do you want me to do that and rebase?

Reply:
Yes please

Reply:
OK. I will send you another pull request with just that today.